### ISSUE TITLE
identify when connectors get into a state where they are not producing messages

### ISSUE THREAD

**Issue by jrhizor** [airbytehq/airbyte#10364]
It's possible to end up in a state where a connector is stuck indefinitely. Here are three classes of problems we've seen recently:
1. The connector gets stuck in an infinite loop. This could be due to a bug, infinite retries for an API, etc.
2. A thread could die (we've seen OOM-killed threads as an example) which isn't recognized by the main thread, which results in the application waiting forever on the killed thread to complete successfully.
3. The connector could be "complete" but something, like a background thread is not cleaned up and the connector gets stuck running even though its work is effectively over.

We need some mechanism for the platform to recognize if a source/destination is in a bad state. Here are a couple of naive approaches and caveats:
- We could have timeouts for messages from the source (this may have to be record messages, since (1) could result in logging messages, but it's hard to detect rate limit backoffs) or the destination (might be hard to calibrate the timeout times for this). 
- We could emit "heartbeat" messages from both the source and the destination. This is pretty implementation-specific though since it may not be able to detect 1) and 2) very well.

So basically for this issue we should have someone on platforms and someone on connectors brainstorm and come up with a path forward.

**Comment by cgardens** [comment]
When did #2 happen? Is there a link to an OC issue or other context?

**Comment by cgardens** [comment]
@jrhizor When did #2 happen? Is there a link to an OC issue or other context?

**Comment by davinchia** [comment]
same question as Charles.

I like the message timeout proposal. I don't think it's that naive. It's simple and covers all the cases. Also lends itself well to the platform/connectors area of responsibilities - connector team is able to define a message timeout per connector, platform ensures a pod doesn't run for too long and passes this back to the team with debug logs if this error happens.

**Comment by pmossman** [comment]
Grooming notes:
- Need to agree on some sort of interface with Connector's team so we can know when a connector is stuck
- On the platform side, we need a failsafe because we can't assume every connector will implement the interface
- Long term discussion with Connectors team probably involves an evolution of the airbyte protocol
- A brainstorm feels necessary for this, some tradeoffs to figure out, a heartbeat is complex and has some issues

**Comment by pmossman** [comment]
Goal for this sprint: Commit to figuring out a plan, but not an implementation

**Comment by jrhizor** [comment]
Created https://github.com/airbytehq/airbyte/issues/11235 to track exposing the stats in the API and https://github.com/airbytehq/airbyte/issues/11236 to handle timeouts for custom connectors.

Closing this issue since we're deciding not to go with heartbeats.
